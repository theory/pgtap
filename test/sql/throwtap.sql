\unset ECHO
\i test/setup.sql

SELECT plan(93);
--SELECT * FROM no_plan();

/****************************************************************************/
-- test throws_ok().
SELECT * FROM check_test(
    throws_ok( 'SELECT * FROM todo_end()', 'P0001', 'todo_end() called without todo_start()', 'whatever' ),
    true,
    'four-argument form',
    'whatever',
    ''
);

SELECT * FROM check_test(
    throws_ok( 'SELECT * FROM todo_end()', 'P0001', 'todo_end() called without todo_start()'),
    true,
    'three-argument errcode',
    'threw P0001: todo_end() called without todo_start()',
    ''
);

SELECT * FROM check_test(
    throws_ok( 'SELECT 1 / 0', '22012' ),
    true,
    'two-argument errcode',
    'threw 22012'
    ''
);

SELECT * FROM check_test(
    throws_ok( 'SELECT * FROM todo_end()', 'todo_end() called without todo_start()', 'whatever'),
    true,
    'three argument errmsg',
    'whatever',
    ''
);

SELECT * FROM check_test(
    throws_ok( 'SELECT * FROM todo_end()', 'todo_end() called without todo_start()'),
    true,
    'two-argument errmsg',
    'threw todo_end() called without todo_start()',
    ''
);

SELECT * FROM check_test(
    throws_ok( 'SELECT 1 / 0' ),
    true,
    'single-argument form',
    'threw an exception',
    ''
);

-- Try using a prepared statement.
PREPARE mytest AS SELECT * FROM todo_end();
SELECT * FROM check_test(
    throws_ok( 'mytest', 'P0001'),
    true,
    'prepared statement & errcode',
    'threw P0001'
    ''
);

SELECT * FROM check_test(
    throws_ok( 'EXECUTE mytest', 'P0001'),
    true,
    'execute & errcode',
    'threw P0001'
    ''
);

-- Check its diagnostics for an invalid error code.
SELECT * FROM check_test(
    throws_ok( 'SELECT * FROM todo_end()', 97212 ),
    false,
    'invalid errcode',
    'threw 97212',
    '      caught: P0001: todo_end() called without todo_start()
      wanted: 97212'
);

SELECT throws_ok( 'SELECT 1 / 0', NULL, NULL, 'throws_ok(1/0, NULL) should work' );

-- Check its diagnostics no error.

SELECT * FROM check_test(
    throws_ok( 'SELECT 1', NULL ),
    false,
    'throws_ok diagnostics',
    'threw an exception',
    '      caught: no exception
      wanted: an exception'
);

/****************************************************************************/
-- test lives_ok().
SELECT lives_ok( 'SELECT 1', 'lives_ok() should work' );

PREPARE livetest AS SELECT 1;
SELECT * FROM check_test(
    lives_ok( 'livetest'),
    true,
    'lives_ok(prepared)'
    '',
    ''
);

SELECT * FROM check_test(
    lives_ok( 'EXECUTE livetest'),
    true,
    'lives_ok(execute)'
    '',
    ''
);

-- Check its diagnostics when there is an exception.
SELECT * FROM check_test(
    lives_ok( 'SELECT * FROM todo_end()' ),
    false,
    'lives_ok failure diagnostics',
    '',
    '        died: P0001: todo_end() called without todo_start()' ||
 E'\n              CONTEXT: SQL statement "SELECT * FROM todo_end()"' ||
 E'\nPL/pgSQL function lives_ok(text,text) line 13 at EXECUTE statement'
);

-- Check the different detail messages
CREATE DOMAIN pg_temp.test_domain boolean NOT NULL;
CREATE TEMP TABLE test_table(b boolean NOT NULL CHECK(b), d pg_temp.test_domain);
SELECT * FROM check_test(
    lives_ok( 'INSERT INTO test_table VALUES(NULL,true)' ),
    false,
    'lives_ok failure diagnostics(schema/table/column)',
    '',
    '        died: 23502: null value in column "b" violates not-null constraint' ||
 E'\n              SCHEMA: ' || ( SELECT nspname FROM pg_namespace WHERE oid = pg_my_temp_schema() ) ||
 E'\n              TABLE: test_table' ||
 E'\n              COLUMN: b' ||
 E'\n              CONTEXT: SQL statement "INSERT INTO test_table VALUES(NULL,true)"' ||
 E'\nPL/pgSQL function lives_ok(text,text) line 13 at EXECUTE statement'
);
SELECT * FROM check_test(
    lives_ok( 'INSERT INTO test_table VALUES(false,true)' ),
    false,
    'lives_ok failure diagnostics(constraint)',
    '',
    '        died: 23514: new row for relation "test_table" violates check constraint "test_table_b_check"' ||
 E'\n              SCHEMA: ' || ( SELECT nspname FROM pg_namespace WHERE oid = pg_my_temp_schema() ) ||
 E'\n              TABLE: test_table' ||
 E'\n              CONSTRAINT: test_table_b_check' ||
 E'\n              CONTEXT: SQL statement "INSERT INTO test_table VALUES(false,true)"' ||
 E'\nPL/pgSQL function lives_ok(text,text) line 13 at EXECUTE statement'
);
SELECT * FROM check_test(
    lives_ok( 'INSERT INTO test_table VALUES(true,NULL)' ),
    false,
    'lives_ok failure diagnostics(constraint)',
    '',
    '        died: 23502: domain test_domain does not allow null values' ||
 E'\n              SCHEMA: ' || ( SELECT nspname FROM pg_namespace WHERE oid = pg_my_temp_schema() ) ||
 E'\n              DATATYPE: test_domain' ||
 E'\n              CONTEXT: SQL statement "INSERT INTO test_table VALUES(true,NULL)"' ||
 E'\nPL/pgSQL function lives_ok(text,text) line 13 at EXECUTE statement'
);

/****************************************************************************/
-- test throws_like().
SELECT * FROM check_test(
    throws_like( 'SELECT * FROM todo_end()', '%end() called without todo%', 'whatever' ),
    true,
    'throws_like(sql, pattern, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    throws_like( 'SELECT * FROM todo_end()', '%end() called without todo%' ),
    true,
    'throws_like(sql, pattern)',
    'Should throw exception like ''%end() called without todo%''',
    ''
);

SELECT * FROM check_test(
    throws_like( 'SELECT * FROM todo_end()', '%huh%', 'whatever' ),
    false,
    'throws_like(sql, pattern, desc) fail',
    'whatever',
    '   error message: ''todo_end() called without todo_start()''
   doesn''t match: ''%huh%'''
);

SELECT * FROM check_test(
    throws_like( 'SELECT 1', '%huh%', 'whatever' ),
    false,
    'throws_like(valid sql, pattern, desc)',
    'whatever',
    '    no exception thrown'
);

/****************************************************************************/
-- test throws_ilike().
SELECT * FROM check_test(
    throws_ilike( 'SELECT * FROM todo_end()', '%END() called without todo%', 'whatever' ),
    true,
    'throws_ilike(sql, pattern, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    throws_ilike( 'SELECT * FROM todo_end()', '%END() called without todo%' ),
    true,
    'throws_ilike(sql, pattern)',
    'Should throw exception like ''%END() called without todo%''',
    ''
);

SELECT * FROM check_test(
    throws_ilike( 'SELECT * FROM todo_end()', '%HUH%', 'whatever' ),
    false,
    'throws_ilike(sql, pattern, desc) fail',
    'whatever',
    '   error message: ''todo_end() called without todo_start()''
   doesn''t match: ''%HUH%'''
);

SELECT * FROM check_test(
    throws_ilike( 'SELECT 1', '%HUH%', 'whatever' ),
    false,
    'throws_ilike(valid sql, pattern, desc)',
    'whatever',
    '    no exception thrown'
);

/****************************************************************************/
-- test throws_matching().
SELECT * FROM check_test(
    throws_matching(
        'SELECT * FROM todo_end()',
        '.*end[(][)] called without todo.+',
        'whatever'
    ),
    true,
    'throws_matching(sql, regex, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    throws_matching(
        'SELECT * FROM todo_end()',
        '.*end[(][)] called without todo.+'
    ),
    true,
    'throws_matching(sql, regex, desc)',
    'Should throw exception matching ''.*end[(][)] called without todo.+''',
    ''
);

SELECT * FROM check_test(
    throws_matching(
        'SELECT * FROM todo_end()',
        'huh.+',
        'whatever'
    ),
    false,
    'throws_matching(sql, regex, desc)',
    'whatever',
    '   error message: ''todo_end() called without todo_start()''
   doesn''t match: ''huh.+'''
);

SELECT * FROM check_test(
    throws_matching(
        'SELECT 1',
        'huh.+',
        'whatever'
    ),
    false,
    'throws_matching(valid sql, regex, desc)',
    'whatever',
    '    no exception thrown'
);

/****************************************************************************/
-- test throws_imatching().
SELECT * FROM check_test(
    throws_imatching(
        'SELECT * FROM todo_end()',
        '.*end[(][)] CALLED without todo.+',
        'whatever'
    ),
    true,
    'throws_imatching(sql, regex, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    throws_imatching(
        'SELECT * FROM todo_end()',
        '.*end[(][)] CALLED without todo.+'
    ),
    true,
    'throws_imatching(sql, regex, desc)',
    'Should throw exception matching ''.*end[(][)] CALLED without todo.+''',
    ''
);

SELECT * FROM check_test(
    throws_imatching(
        'SELECT * FROM todo_end()',
        'HUH.+',
        'whatever'
    ),
    false,
    'throws_imatching(sql, regex, desc)',
    'whatever',
    '   error message: ''todo_end() called without todo_start()''
   doesn''t match: ''HUH.+'''
);

SELECT * FROM check_test(
    throws_imatching(
        'SELECT 1',
        'HUH.+',
        'whatever'
    ),
    false,
    'throws_imatching(valid sql, regex, desc)',
    'whatever',
    '    no exception thrown'
);

/****************************************************************************/
-- Finish the tests and clean up.
SELECT * FROM finish();
ROLLBACK;
