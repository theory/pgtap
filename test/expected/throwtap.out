\unset ECHO
1..93
ok 1 - four-argument form should pass
ok 2 - four-argument form should have the proper description
ok 3 - four-argument form should have the proper diagnostics
ok 4 - three-argument errcode should pass
ok 5 - three-argument errcode should have the proper description
ok 6 - three-argument errcode should have the proper diagnostics
ok 7 - two-argument errcode should pass
ok 8 - two-argument errcode should have the proper description
ok 9 - three argument errmsg should pass
ok 10 - three argument errmsg should have the proper description
ok 11 - three argument errmsg should have the proper diagnostics
ok 12 - two-argument errmsg should pass
ok 13 - two-argument errmsg should have the proper description
ok 14 - two-argument errmsg should have the proper diagnostics
ok 15 - single-argument form should pass
ok 16 - single-argument form should have the proper description
ok 17 - single-argument form should have the proper diagnostics
ok 18 - prepared statement & errcode should pass
ok 19 - prepared statement & errcode should have the proper description
ok 20 - execute & errcode should pass
ok 21 - execute & errcode should have the proper description
ok 22 - invalid errcode should fail
ok 23 - invalid errcode should have the proper description
ok 24 - invalid errcode should have the proper diagnostics
ok 25 - throws_ok(1/0, NULL) should work
ok 26 - throws_ok diagnostics should fail
ok 27 - throws_ok diagnostics should have the proper description
ok 28 - throws_ok diagnostics should have the proper diagnostics
ok 29 - lives_ok() should work
ok 30 - lives_ok(prepared) should pass
ok 31 - lives_ok(prepared) should have the proper description
ok 32 - lives_ok(execute) should pass
ok 33 - lives_ok(execute) should have the proper description
ok 34 - lives_ok failure diagnostics should fail
ok 35 - lives_ok failure diagnostics should have the proper description
ok 36 - lives_ok failure diagnostics should have the proper diagnostics
ok 37 - lives_ok failure diagnostics(schema/table/column) should fail
ok 38 - lives_ok failure diagnostics(schema/table/column) should have the proper description
ok 39 - lives_ok failure diagnostics(schema/table/column) should have the proper diagnostics
ok 40 - lives_ok failure diagnostics(constraint) should fail
ok 41 - lives_ok failure diagnostics(constraint) should have the proper description
ok 42 - lives_ok failure diagnostics(constraint) should have the proper diagnostics
ok 43 - lives_ok failure diagnostics(constraint) should fail
ok 44 - lives_ok failure diagnostics(constraint) should have the proper description
ok 45 - lives_ok failure diagnostics(constraint) should have the proper diagnostics
ok 46 - throws_like(sql, pattern, desc) should pass
ok 47 - throws_like(sql, pattern, desc) should have the proper description
ok 48 - throws_like(sql, pattern, desc) should have the proper diagnostics
ok 49 - throws_like(sql, pattern) should pass
ok 50 - throws_like(sql, pattern) should have the proper description
ok 51 - throws_like(sql, pattern) should have the proper diagnostics
ok 52 - throws_like(sql, pattern, desc) fail should fail
ok 53 - throws_like(sql, pattern, desc) fail should have the proper description
ok 54 - throws_like(sql, pattern, desc) fail should have the proper diagnostics
ok 55 - throws_like(valid sql, pattern, desc) should fail
ok 56 - throws_like(valid sql, pattern, desc) should have the proper description
ok 57 - throws_like(valid sql, pattern, desc) should have the proper diagnostics
ok 58 - throws_ilike(sql, pattern, desc) should pass
ok 59 - throws_ilike(sql, pattern, desc) should have the proper description
ok 60 - throws_ilike(sql, pattern, desc) should have the proper diagnostics
ok 61 - throws_ilike(sql, pattern) should pass
ok 62 - throws_ilike(sql, pattern) should have the proper description
ok 63 - throws_ilike(sql, pattern) should have the proper diagnostics
ok 64 - throws_ilike(sql, pattern, desc) fail should fail
ok 65 - throws_ilike(sql, pattern, desc) fail should have the proper description
ok 66 - throws_ilike(sql, pattern, desc) fail should have the proper diagnostics
ok 67 - throws_ilike(valid sql, pattern, desc) should fail
ok 68 - throws_ilike(valid sql, pattern, desc) should have the proper description
ok 69 - throws_ilike(valid sql, pattern, desc) should have the proper diagnostics
ok 70 - throws_matching(sql, regex, desc) should pass
ok 71 - throws_matching(sql, regex, desc) should have the proper description
ok 72 - throws_matching(sql, regex, desc) should have the proper diagnostics
ok 73 - throws_matching(sql, regex, desc) should pass
ok 74 - throws_matching(sql, regex, desc) should have the proper description
ok 75 - throws_matching(sql, regex, desc) should have the proper diagnostics
ok 76 - throws_matching(sql, regex, desc) should fail
ok 77 - throws_matching(sql, regex, desc) should have the proper description
ok 78 - throws_matching(sql, regex, desc) should have the proper diagnostics
ok 79 - throws_matching(valid sql, regex, desc) should fail
ok 80 - throws_matching(valid sql, regex, desc) should have the proper description
ok 81 - throws_matching(valid sql, regex, desc) should have the proper diagnostics
ok 82 - throws_imatching(sql, regex, desc) should pass
ok 83 - throws_imatching(sql, regex, desc) should have the proper description
ok 84 - throws_imatching(sql, regex, desc) should have the proper diagnostics
ok 85 - throws_imatching(sql, regex, desc) should pass
ok 86 - throws_imatching(sql, regex, desc) should have the proper description
ok 87 - throws_imatching(sql, regex, desc) should have the proper diagnostics
ok 88 - throws_imatching(sql, regex, desc) should fail
ok 89 - throws_imatching(sql, regex, desc) should have the proper description
ok 90 - throws_imatching(sql, regex, desc) should have the proper diagnostics
ok 91 - throws_imatching(valid sql, regex, desc) should fail
ok 92 - throws_imatching(valid sql, regex, desc) should have the proper description
ok 93 - throws_imatching(valid sql, regex, desc) should have the proper diagnostics
